**Due Date: Monday, January 29th, 11:59pm**

- Save a copy of the notebook to your Google Drive. You can do that by going to the menu and clicking `File` > `Save` > `SAVE A COPY IN DRIVE`.
- Fill out the missing parts and run the code modules.
- Answer the questions (if any) in a separate document or by adding a new `Text` block inside the Colab.
- Download the solved notebook by going to the menu and clicking `File` > `Download .ipynb`.
- Make sure the downloaded version is showing your solutions.
- Upload your solutions to BruinLearn (under "Colab Assignment #2: Hypothesis Testing" for Week 3).

In [1]:
import numpy as np
from scipy import stats

np.random.seed(0)

Assume `y_1`, `y_2`, ..., `y_6` show the outcome of coin flips for six different coins.

In [2]:
y_1 = np.random.binomial(1, 0.5, 50)
y_2 = np.random.binomial(1, 0.5, 30)
y_3 = np.random.binomial(1, 0.5, 40)
y_4 = np.random.binomial(1, 0.5, 10)
y_5 = np.random.binomial(1, 0.4, 30)
y_6 = np.random.binomial(1, 0.8, 50)

## One group

Based on the observed data, test the hypothesis that the first coin is fair. The null hypothesis for this test is $H_0: p_1 = 0.5$. Compute the p-value of this test. Do you reject the null hypothesis based on the result of this test? If not, does that mean that $p_1 = 0.5$?



In [7]:
# TODO: Assign the right value to p_val

mu_0 = 0.5
t_stat, p_val = stats.ttest_1samp(y_1, mu_0)

print(t_stat)
print(p_val)

# END OF TODO

print("The p-value for the test is", p_val)

1.4288690166235203
0.15938605197461647
The p-value for the test is 0.15938605197461647


Since the p-value is 0.159, which is greater than the 0.05 significance level, we can conclude that the result is not statistically significant meaning it is not strong enough evidence to reject the null hypothesis.

Answer the same questions for the fifth coin. This time, the null hypothesis is $H_0: p_5 = 0.5$.



In [8]:
# TODO: Assign the right value to p_val

mu_0 = 0.5
t_stat, p_val = stats.ttest_1samp(y_5, mu_0)

print(t_stat)
print(p_val)

# END OF TODO

print("The p-value for the test is", p_val)

-0.3598114585644621
0.7215952675845513
The p-value for the test is 0.7215952675845513


Since the p-value is 0.722, which is greater than the 0.05 significance level, we can conclude that the result is not statistically significant meaning it is not strong enough evidence to reject the null hypothesis.

Answer the same questions for the sixth coin. This time, the null hypothesis is $H_0: p_6 = 0.5$.

In [9]:
# TODO: Assign the right value to p_val

mu_0 = 0.5
t_stat, p_val = stats.ttest_1samp(y_6, mu_0)

print(t_stat)
print(p_val)

# END OF TODO

print("The p-value for the test is", p_val)

0.5618006477335901
0.5768126908825979
The p-value for the test is 0.5768126908825979


Since the p-value is 0.578, which is greater than the 0.05 significance level, we can conclude that the result is not statistically significant meaning it is not strong enough evidence to reject the null hypothesis.

## Two groups

Test if the bias of the fifth coin and the sixth coin are equal to each other. The null hypothesis is $H_0: p_5 = p_6$. Compute the p-value for this test.

In [11]:
# TODO: Assign the right value to p_val

t_stat, p_val = stats.ttest_ind(y_5, y_6, equal_var = True)

print(t_stat)
print(p_val)

# END OF TODO

print("The p-value for the test is", p_val)

-0.05746065660284369
0.9543251930320562
The p-value for the test is 0.9543251930320562


## k groups

Test the hypothesis that all the six coins have the same bias. The null hypothesis is $H_0: p_1 = p_2 = \dots = p_6$. Compute the p-value for this test. Can you reject the null hypothesis based on this p-value?

In [12]:
# TODO: Assign the right value to p_val

f_stat, p_val = stats.f_oneway(y_1, y_2, y_3, y_4, y_5, y_6)
print(f_stat)
print(p_val)

# END OF TODO

print("The p-value for the test is", p_val)

1.5882121770435256
0.16485237297841546
The p-value for the test is 0.16485237297841546


Since the p-value is 0.165, which is greater than the 0.05 significance level, we can conclude that the result is not statistically significant meaning it is not strong enough evidence to reject the null hypothesis.

Test the hypothesis that all the first four coins have the same bias. The null hypothesis is $H_0: p_1 = p_2 = p_3 = p_4$. Compute the p-value for this test. Can you reject the null hypothesis based on this p-value?

In [13]:
# TODO: Assign the right value to p_val

f_stat, p_val = stats.f_oneway(y_1, y_2, y_3, y_4)
print(f_stat)
print(p_val)

# END OF TODO

print("The p-value for the test is", p_val)

2.0907031363664705
0.1047599747354836
The p-value for the test is 0.1047599747354836


Since the p-value is 0.105, which is greater than the 0.05 significance level, we can conclude that the result is not statistically significant meaning it is not strong enough evidence to reject the null hypothesis.

This time, let's do six pairwise two-sample t-tests between the first four coins. Meaning, let's test the null hypothesis $H_0^{(i,j)}: p_i = p_j$ for each $(i,j)$ pair where $1 \leq i \leq 4$, $1 \leq j \leq 4$, and $i \neq j$). What is the smallest p-value you get from these six tests?

In [18]:
# TODO: Assign the right value to smallest_p_val
y = [y_1, y_2, y_3, y_4]

smallest_p_val = 1

for i in range(0, 3):
    t_stat, p_val = stats.ttest_ind(y[i], y[i+1], equal_var = True)
    smallest_p_val = min(p_val, smallest_p_val)
# END OF TODO

print("Smallest p-value in pair-wise t-tests is: ", smallest_p_val)

Smallest p-value in pair-wise t-tests is:  0.020754896218989474


If we test $H_0: p_1 = p_2 = p_3 = p_4$ by doing pairwise two-sample t-tests and rejecting $H_0: p_1 = p_2 = p_3 = p_4$ if we could reject any one of these pairwise tests, will we reject $H_0$? If so, is that the right decision?

Since the p-value is 0.02, which is less than the 0.05 significance level, we can conclude that the result is statistically significant meaning it is strong enough evidence to reject the null hypothesis.